File: 844-backspace-string-compare/844-backspace-string-compare.cpp
class Solution {
public:
    bool backspaceCompare(string s, string t) {
        stack<char> sdash;
        stack<char> tdash;
        
        for(char ch : s) {
            if(ch == '#') {
                if(sdash.size() > 0) {
                    sdash.pop();
                }
            } else {
                sdash.push(ch);
            }
        }
        
        for(char ch : t) {
            if(ch == '#') {
                if(tdash.size() > 0) {
                    tdash.pop();
                }
            } else {
                tdash.push(ch);
            }
        }
        
        if(sdash.size() != tdash.size()) {
            return false;
        }
        
        while(sdash.size() != 0) {
            if(sdash.top() != tdash.top()) {
                return false;
            } else {
                sdash.pop();
                tdash.pop();
            }
        }
        
        return true;
    }
};

File: 844-backspace-string-compare/844-backspace-string-compare.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        int i = s.length() - 1, j = t.length() - 1;
        while(i >= 0 || j >= 0) {
            int backSpace = 0;
            while(i >= 0) {
                if(s.charAt(i) == '#') {
                    i--;
                    backSpace++;
                } else if(backSpace > 0) {
                    i--;
                    backSpace--;
                } else {
                    break;
                }
            }
            
            backSpace = 0;
            while(j >= 0) {
                if(t.charAt(j) == '#') {
                    j--;
                    backSpace++;
                } else if(backSpace > 0) {
                    j--;
                    backSpace--;
                } else {
                    break;
                }
            }
            
            if(i < 0 && j < 0) {
                return true;
            }
            
            if(i < 0 || j < 0) {
                return false;
            }
            
            if(s.charAt(i) != t.charAt(j)) {
                return false;
            }
            
            i--;
            j--;
        }
        
        return true;
    }
}